Fixes #3553, Fixes #3790 AZ Person manageable using CAS account login, Dedicated Permission for Publication Curation

tadean opened the pull request:
## Description
This PR adds a new role known (so far) as `Self-Service Editor`. The intended purpose of this role is to edit a Person belonging in some way to the user.

A new proposed permission is added; `edit matching netid content`. This permission grants edit permission to nodes which have a netID field which matches the user's externalauth authmap (CAS, in our case)

This is useful in cases of integrations where Person content has been migrated in bulk and the user does not own the Person node in question (and existing administrators do not have to establish the ownership relationship for existing content).

For review/discussion

- What should we call this role?
- What permissions should this role have?
- This role is a departure from our additive approach to permissions where we do not normally allow permissions to overlap between roles. This is because this role is entirely subtractive in nature, we do not want it to have all editor permissions. We should discuss thoughts about this.
- @trackleft proposes adding a local task for the user that prompts them to edit their Person when logging in. This seems like a nice feature to have.

### Release notes
```
 A new Profile Editor role has been added to Quickstart. This role gains permission to edit Person nodes where the user is assigned as the author, or Person nodes that have the same NetID as the user's login.
 
 The role also has the ability to change the order and visibility of publications shown on the Publication block on their Person page.
```

## Related issues
Closes #3553 
Closes #3790 

## How to test

- import demo content
- assign your netid to a person
- create a CAS account for yourself on the site
- grant ownership of a (different) person to your account
- Log in as your CAS account
- verify you have permission to edit both the person you are author of, and the person that shares your netid

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

tadean commented:
> Alternatively, could this person self editing be achieved with assigning the node author as the relevant user? 

This PR does this also. The role has permission `edit own az_person content`

The purpose of the NetID/CAS permission is to support the use case where a site has a large amount of users created by integrations and the site administrator does not want to individually manage node ownership to all the users.

tadean commented:
> PTYS uses Content Access and ACL modules to allow self-service editing of profiles.

In your experience on PTYS, which permissions do you grant to them beyond the node access? (with regard specifically to  editing-related permissions like text formats/administration themes/etc.)

tadean commented:
Thanks for this list, very helpful... good idea on contextual links! I agree the admin theme would be pretty useful...it seems like there's probably some unanticipated weirdness using the edit forms via the front end theme.

joeparsons commented on modules/custom/az_core/az_core.module:
```diff
@@ -437,6 +439,38 @@ function az_core_form_taxonomy_overview_terms_alter(&$form, FormStateInterface $
+        try {
+          // Look up user mapping for CAS.
+          $auth = \Drupal::service('externalauth.authmap')->get($account->id(), 'cas');
+          \Drupal::logger('netid_auth')->notice($auth);
```
```suggestion
```
Is this notice intended to be a permanent addition?  (Wasn't sure if it was for debuging)

tadean commented:
@camikazegreen suggests ability to curate publication order attached to their person page

danahertzberg commented:
Here are some details on a customization I did for the Neuroscience site to allow students to edit their own person listings.
https://neuroscience.arizona.edu/students/student-directory

* Modified the az_person content type
   * Added field_neuro_advisor
   * Added field_neuro_current_employer
   * Added field_neuro_current_lab
   * Added field_neuro_dissertation_advisor
   * Added field_neuro_dissertation_title
   * Added field_neuro_dissertation_year
   * Added field_neuro_minors
   * Added field_neuro_research_interest
   * Added field_neuro_rotation_labs
   * Modified existing content type displays
   * Added new content type displays
* Modified AZ Person view
* Added new Student role with the following permissions
   * Use contextual links
   * Access Draggableviews
   * See environment indicator
   * Access the Files overview page
   * Delete own files
   * Use the [Standard](https://neuroscience.arizona.edu/admin/config/content/formats/manage/az_standard) text format
   * Use help pages
   * Document: Create new media
   * Image: Create new media
   * Document: Edit own media
   * Image: Edit own media
   * Create media
   * Delete own media
   * Update own media
   * View all media revisions
   * View own unpublished media
   * Person: Create new content
   * Person: Edit own content
   * Person: Revert revisions
   * Person: View revisions

tadean commented:
Added functionality that adds a user toolbar link for editing the associated az_person.

tadean commented:
Added permissions for self-service editors to curate their own publications.

joeparsons commented on modules/custom/az_core/src/AZUserToolbarLinkBuilder.php:
Dumb question:  Wouldn't it be possible to omit the constructor and class properties set by the constructor and just defer to the parent class for those?

joeparsons replied:
Never mind.  I got confused by the externalauth module name and didn't realize this was a AZ Person / CAS specific implementation different than the parent class.

danahertzberg commented:
The term `Self-service editor` for permissions makes sense to me.

My vote for menu options.

- View my web page
- Edit my web page
- View user account
- Edit user account

Would it make sense to move the `View/edit my web page` links under Content rather than the User? I'm good either way.

tadean commented:
- Changed service implementation to use `parent:` at @joeparsons suggestion to make clear which arguments are derived
- Added @danahertzberg 's suggested link titles
- Changed service implementation to decorate the parent service so that an alter is not necessary

Should this role not have access to the analytics dashboard? @joshuasosa suggested this is done on sites in their portfolio that have similar functionality

tadean commented on modules/custom/az_publication/config/install/views.view.az_publications_curate.yml:
```diff
@@ -846,7 +846,8 @@ display:
           required: true
       display_description: ''
       display_extenders: {  }
-      path: admin/az-publication/curation/%node
```
This change is made because the new proposed role should have access to curate publications, but does not have access to `Use the administration pages` permission. This is different from the permission `view the administration theme`, which the role does have.

joeparsons commented on config/install/user.role.az_self_service_editor.yml:
```diff
@@ -0,0 +1,25 @@
+status: true
+dependencies: {  }
+id: az_self_service_editor
+label: 'Self-service editor'
```
```suggestion
label: 'Self-service profile editor'
```

This is the best suggestion I can think for for a more specific role name.

tadean commented:
Incorporated feedback for changing order of roles.

tadean commented:
> This is the best suggestion I can think for for a more specific role name.

@joeparsons I've incorporated this suggestion for review. I've also been wondering if a very minimal version would be explanatory enough, like simply "Self editor" (or something), for purposes of keeping the role name concise.

joeparsons commented:
> > This is the best suggestion I can think for for a more specific role name.
> 
> @joeparsons I've incorporated this suggestion for review. I've also been wondering if a very minimal version would be explanatory enough, like simply "Self editor" (or something), for purposes of keeping the role name concise.

Oh, I think "Self editor" sounds promising.

danahertzberg commented:
> Oh, I think "Self editor" sounds promising.

"Self-service editor"? I like the self-service term personally.
